cclib/method/bader.py

```diff
@@ -0,0 +1,223 @@
+        self.fragresults = None
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+
+        # First obtain charge densities on the grid
+        if len(self.data.mocoeffs) == 1:
+            self.chgdensity = electrondensity_spin(
+                self.data, self.volume, [self.data.mocoeffs[0][: self.data.homos[0]]]
+            )
+            self.chgdensity.data *= 2
+        else:
+            self.chgdensity = electrondensity_spin(
+                self.data,
+                self.volume,
+                [
+                    self.data.mocoeffs[0][: self.data.homos[0]],
+                    self.data.mocoeffs[1][: self.data.homos[1]],
+                ],
+            )
+
+        # Assign each grid point to Bader areas
+        self.fragresults = numpy.zeros(self.chgdensity.data.shape, "d")
+        nextIndex = 1
+
+        self.logger.info("Partitioning space into Bader areas.")
+
+        # Exclude the outermost positions
+        for xwalker in range(1, len(self.chgdensity.data) - 1):
+            for ywalker in range(1, len(self.chgdensity.data[0]) - 1):
+                for zwalker in range(1, len(self.chgdensity.data[0][0]) - 1):
+                    (xindex, yindex, zindex) = (xwalker, ywalker, zwalker)
+
+                    if self.fragresults[xindex, yindex, zindex] != 0:
+                        # index has already been assigned for this grid point
+                        continue
+                    else:
+                        listcoord = []
+                        loopcondition = True
+
+                        while loopcondition:
+                            # Here, `delta_rho` corresponds to equation 2,
+                            # and `grad_rho_dot_r` corresponds to equation 1 in the aforementioned
+                            # paper (doi:10.1016/j.commatsci.2005.04.010)
+                            delta_rho = (
+                                self.chgdensity.data[
+                                    xindex - 1 : xindex + 2,
+                                    yindex - 1 : yindex + 2,
+                                    zindex - 1 : zindex + 2,
+                                ]
+                                - self.chgdensity.data[xindex, yindex, zindex]
+                            )
+                            grad_rho_dot_r = delta_rho / _griddist
```

Comment:
Looking at it again, I'm surprised this works since `_griddist` isn't an `ndarray`. Can you try making it one and make sure the results don't change?

Reply:
I think this worked because division operator for Python on lists are elementwise divide as well? I've changed this into ndarray for consistency. As a side note, I forced pushed the commit by mistake -- this became some sort of a reflex :0